go_router/lib/src/route_data.dart

```diff
@@ -0,0 +1,120 @@
+import 'package:flutter/widgets.dart';
+import 'package:meta/meta.dart';
+import 'package:meta/meta_meta.dart';
+
+import 'go_route.dart';
+import 'go_router_state.dart';
+
+/// Baseclass for types that support
+/// [typed routing](https://gorouter.dev/typed-routing).
+///
+/// Subclasses must implement [build]. They can also, optionally override
+/// [redirect].
+abstract class GoRouteData {
+  /// Allows subclasses to have `const` constructors.
+  ///
+  /// [GoRouteData] is abstract and cannot be instantiated directly.
+  const GoRouteData();
+
+  /// Creates the [Widget] associated with `this` instance when the
+  /// corresponding route is navigated to.
+  ///
+  /// Corresponds to [GoRoute.builder].
+  Widget build(BuildContext context) => throw UnimplementedError(
+        'One of `build` or `buildPage` must be implemented.',
+      );
+
+  /// By default, returns a [Page] instance that is ignored, causing a default
+  /// [Page] implementation to be used with the results of [build].
+  ///
+  /// Subclasses can override this function to provide a custom [Page].
+  ///
+  /// Corresponds to [GoRoute.pageBuilder].
+  Page<void> buildPage(BuildContext context) => const NoOpPage();
+
+  /// When overridden in a subclass, allows specifying a path to redirect to.
+  ///
+  /// Corresponds to [GoRoute.redirect].
+  String? redirect() => null;
+
+  /// A helper function used by generated code. Should not be used directly.
+  static String $location(String path, {Map<String, String>? queryParams}) =>
+      Uri.parse(path)
+          .replace(
+            queryParameters:
+                // Avoid `?` in generated location if `queryParams` is empty
+                queryParams == null || queryParams.isEmpty ? null : queryParams,
+          )
+          .toString();
+
+  /// A helper function used by generated code. Should not be used directly.
+  static GoRoute $route<T extends GoRouteData>({
+    required String path,
+    required T Function(GoRouterState) factory,
+    List<GoRoute> routes = const [],
+  }) {
+    T factoryImpl(GoRouterState state) {
+      final extra = state.extra;
+
+      if (extra is T) {
+        return extra;
+      }
+
+      return (_stateObjectExpando[state] ??= factory(state)) as T;
```

Comment:
It seems that `extra` isn't always populated in some of these flows. Maybe around redirects? Might be worth investigating @csells

Reply:
Can you provide me a minimal repro?

Reply:
The minimal repo is to put in print debugging and see the code paths that come through here where there is no extra.

Guessing it's when redirects are calculated.

Reply:
I'm asking for a set of routes and a set of user actions that produces a null extra parameter.

Reply:
Put a breakpoint on line 63 and run the example in debug mode...that'll do it!